UMAPs faceted by highly variable genes

allyhawkins opened the pull request:
Closes #132 

Here I added a section at the end of the report showing a grid of UMAPs with the most highly expressed genes. I started with 12 genes, but we can change that based on what others think of the plots. I did some adjusting to how the plot looks, but I'm not sold on any of it so if others have suggestions on how to make it pretty, I'm open to those! 

I also was wondering if we should make the number of genes we select be a parameter, or do we want to just stick with a number? 

The other thing I did here was utilize the gene symbols that we store in the `rowData` of our objects. I could have plotted ensembl gene ID's as the names, but I thought only Josh would understand those so our users might appreciate actually knowing what the genes were. 

The only other thing I didn't do was make sure that the plots are showing up in order of most variation to least variation among the top 12 genes that I grabbed. If we do want to ensure that happens there's a bit more work that we have to do in creating the plot. 

Here's an example of what the plot looks like right now after rendering in the report: 
![image](https://user-images.githubusercontent.com/54039191/198377190-13ae6a2e-a543-427b-859d-8830814778be.png)

jashapiro commented on inst/rmd/qc_report.rmd:
```diff
@@ -420,7 +423,83 @@ if(has_processed){
+  if(has_var_genes){
+    
+    # select top genes to plot 
+    top_genes <- processed_meta$highly_variable_genes %>%
```
A question that may inform later questions... are the hvgs in the metadata in order from most variable to least? If so, we might want to make this an ordered factor to preserve that. If not, we should change so it is, or this might not be the "top" genes.

allyhawkins replied:
Yes, when I was playing around with this before I checked that they are in order of variance and the first gene that's returned is the one with the highest variation.

jashapiro reviewed (commented):
Just the one question to start... on my phone, but I thought I'd ask.

jashapiro commented on inst/rmd/qc_report.rmd:
```diff
@@ -420,7 +423,83 @@ if(has_processed){
+    
+    # grab rowdata as data frame to later combine with gene expression data
+    # rowdata contains the mapped gene symbol so can use that to label plots instead of ensembl gene id from rownames 
+    rowdata_df <- rowData(processed_sce) %>%
+      as.data.frame() %>%
+      tibble::rownames_to_column("ensembl_id") %>%
+      dplyr::select(ensembl_id, gene_symbol) %>%
+      dplyr::filter(ensembl_id %in% top_genes)
```
I _think_ this will put everything in the correct order.

Note I also changed the pipe operator to match the earlier part of the notebook. This notebook also doesn't use dplyr prefixes in general. We don't load many packages, but I tend to feel that `dplyr` is okay in certain contexts... and I guess I had decided this was one of them.

```suggestion
    rowdata_df <- rowData(processed_sce) |>
      as.data.frame() |>
      tibble::rownames_to_column("ensembl_id") |>
      select(ensembl_id, gene_symbol) |>
      filter(ensembl_id %in% top_genes) |>
      mutate(ensembl_id = factor(ensembl_id, levels = top_genes)) |>
      arrange(ensembl_id) |>
      mutate(gene_symbol = factor(gene_symbol, levels = gene_symbol))
      
```

allyhawkins commented:
Thanks for the suggestion! I went ahead and implemented the change to factors and reordered and that switched the order of the plots. I also removed the magrittr pipe and put the base R pipe. I had forgotten that we had been using that here. Should be ready for another look!

jashapiro commented on inst/rmd/qc_report.rmd:
```diff
@@ -420,7 +423,87 @@ if(has_processed){
 The raw counts from all remaining cells are then normalized prior to selection of highly variable genes and dimensionality reduction. 
-Here we show the UMAP (Uniform Manifold Approximation and Projection) embeddings for each cell, coloring each cell by the total number of genes detected per cell.
+The above plot shows the UMAP (Uniform Manifold Approximation and Projection) embeddings for each cell, coloring each cell by the total number of genes detected per cell.
+The below plots show the same UMAP embeddings, coloring each cell by the gene expression of the gene indicated in the label for each panel. The genes chosen for plotting are the top 12 highly variable genes identified in the library.
```
Simplifying the wording here a bit.
```suggestion
The plots below show the same UMAP embeddings, coloring each cell by the expression level of the labeled gene.
The genes chosen for plotting are the 12 most variable genes identified in the library.
```

jashapiro commented on inst/rmd/qc_report.rmd:
````diff
@@ -420,7 +423,87 @@ if(has_processed){
+The below plots show the same UMAP embeddings, coloring each cell by the gene expression of the gene indicated in the label for each panel. The genes chosen for plotting are the top 12 highly variable genes identified in the library.
+
+```{r message=FALSE}
+if(has_processed){
+  
+  if(has_var_genes){
````
I think we only need `if(has_var_genes){` here? 

In the current setup, if `has_var_genes` is False, the `ggplot` will fail as `coldata_df` wasn't yet set up, correct?

For this to work, we will want to initialize `has_var_genes <- FALSE` before line 387, or make the line something like: 
``` 
if(has_processed & has_var_genes){
```

jashapiro commented on inst/rmd/qc_report.rmd:
```diff
@@ -420,7 +423,87 @@ if(has_processed){
+    facet_wrap(~ gene_symbol) +
+    scale_color_viridis_c() +
+    labs(
+      color = "Log-normalized \ngene expression"
```
I don't think we need to split this onto two lines, given that we are putting the gradient horizontal at the top (or bottom)
```suggestion
      color = "Log-normalized gene expression"
```

jashapiro commented on inst/rmd/qc_report.rmd:
```diff
@@ -420,7 +423,87 @@ if(has_processed){
+    labs(
+      color = "Log-normalized \ngene expression"
+    ) +
+    theme(legend.position = "top", 
```
I might like the bottom better?
```suggestion
    theme(legend.position = "bottom", 
```

jashapiro commented on inst/rmd/qc_report.rmd:
```diff
@@ -420,7 +423,87 @@ if(has_processed){
+          axis.title = element_text(size = 8, color = "black"),
+          legend.title = element_text(size = 8),
+          legend.text = element_text(size = 6)) + 
+    guides(colour = guide_colorbar(title.position = "top", title.hjust = 0.5))
```
If putting the legend on the bottom, probably the bottom for the title too...
```suggestion
    guides(colour = guide_colorbar(title.position = "bottom", title.hjust = 0.5))
```

jashapiro commented on inst/rmd/qc_report.rmd:
```diff
@@ -420,7 +423,87 @@ if(has_processed){
+          axis.text = element_text(size = 8, color = "black"),
+          axis.title = element_text(size = 8, color = "black"),
+          legend.title = element_text(size = 8),
+          legend.text = element_text(size = 6)) + 
```
I probably wouldn't make these quite so small... 8 pt is already pretty small!

I wonder if we want to play with `strip.text` and `strip.background`. I think the grey background is okay, but we might go a bit smaller, or match the axis text, or no border on the background?

I might try:
```
strip.background = element_rect(fill = 'transparent'),
strip.text = element_text(size = 8) 
```

But not something to spend much time on.

jashapiro reviewed (approved):
This looks good to me, with one logic question and a couple of aesthetic suggestions, but I'm really not too concerned about the latter, so I am approving!
